### ISSUE TITLE
[MPS] scaled_dot_product_attention with is_causal=True leaks future tokens in float16/bfloat16

### ISSUE THREAD

**Issue by songdc98** [pytorch/pytorch#195910]
### 🐛 Describe the bug

**Reporter:** Dachuan Song  
**Affiliation:** Department of Electrical and Computer Engineering, George Mason University  
**Contact:** dsong25@gmu.edu

Reporter context: I reproduced this issue on my Apple Silicon Mac and verified the claims below against the reproduction scripts and stored measurement files. I am reporting it because the failure is silent and materially changes model output.

AI assistance disclosure: Codex helped organize and edit the quoted report below from my reproduction code and measurements. I reviewed the technical content before submission.

> ## Summary
>
> On the MPS backend, `torch.nn.functional.scaled_dot_product_attention(..., is_causal=True)` can apply the causal mask once per group of four query positions in `float16` and `bfloat16`. Three positions in each group can therefore attend to future keys. The call returns plausible tensors without a warning, error, or NaN.
>
> ## Minimal reproduction
>
> ```python
> import torch
> import torch.nn.functional as F
>
> B, H, L, D = 1, 4, 64, 64
> torch.manual_seed(0)
> q, k, v = (torch.randn(B, H, L, D) for _ in range(3))
>
> for dtype in (torch.bfloat16, torch.float16):
>     qm, km, vm = (x.to("mps", dtype) for x in (q, k, v))
>     mask = torch.ones(L, L, dtype=torch.bool, device="mps").tril()
>     with torch.no_grad():
>         causal = F.scaled_dot_product_attention(qm, km, vm, is_causal=True)
>         explicit = F.scaled_dot_product_attention(qm, km, vm, attn_mask=mask)
>     print(dtype, (causal - explicit).abs().max().item())
> ```
>
> ## Expected behavior
>
> Query position `i` should attend only to keys `0..i`. The `is_causal=True` and explicit Boolean `attn_mask` paths should agree within normal half-precision rounding error.
>
> ## Actual behavior
>
> Query position `i` can attend through key `(i//4)*4+3`. At query position 0, 86% of the recovered attention weight falls on future tokens. Across the tested boundary grid, all 108 affected cases match this block-of-four mask.
>
> At sequence length 64, the maximum absolute errors against a CPU float64 reference were:
>
> | dtype | `is_causal=True` | explicit Boolean mask |
> |---|---:|---:|
> | `bfloat16` | 2.8655 | 0.0116 |
> | `float16` | 2.8709 | 0.0012 |
>
> `float32` is unaffected. Passing the same causal mask explicitly through `attn_mask` produces the expected result in half precision.
>
> ## Additional evidence and workaround
>
> The repository contains the full 30-line reproduction, raw JSON measurements, boundary tests, mask recovery, model-level results, figures, and a drop-in workaround:
>
> https://github.com/songdc98/apple-silicon-attention-bug
>
> Suggested label: `module: mps`

### Versions

PyTorch version: 2.11.0
OS: macOS 27.0, build 26A5425a
Python version: 3.13.11
Architecture: arm64
MPS available: yes
MPS built: yes
Hardware: 2026 16-inch MacBook Pro; Apple M5 Max; 18-core CPU; 40-core GPU; 128 GB unified memory
Metal support: Metal 4

cc @kulinseth @malfet @DenisVieriu97 @jhavukainen @aditvenk @Isalia20 @drisspg @liangel-02 @howardzhang-cv

**Comment by Isalia20** [comment]
Thanks for the report. I think the issue is the torch version you are using (2.11). Can you update pytorch to the latest version? On latest main I get:
```
torch.bfloat16 0.0
torch.float16 0.0
```
On 2.13 I also get:
```
torch.bfloat16 0.0
torch.float16 0.0
```

On 2.11 I can reproduce the issue.

**Comment by songdc98** [comment]
Confirmed: upgrading fixes it.

I tested four versions in clean venvs on the same machine (M5 Max, macOS 27.0,
Python 3.13) with the repro.py from the report:

  2.11.0  AFFECTED      bf16 is_causal err 4.1389
  2.12.1  AFFECTED      bf16 is_causal err 4.1389
  2.13.0  not affected  bf16 is_causal err 0.0110
  2.14.0  not affected  bf16 is_causal err 0.0085

So the fix landed between 2.12.1 and 2.13.0.
End-to-end on Qwen2.5-1.5B-Instruct, max logit deviation from a CPU fp32 reference for the bf16 + SDPA path: 13.49 on 2.11.0 (0/10 greedy traces match the reference) vs 0.72 on 2.14.0 (8/10 match, same as the bf16 eager path).

Closing from my side. Thanks for looking at it.
